packages/gluestick/shared/containers/Root.js

```
@@ -73,28 +73,39 @@ export default class Root extends Component<DefaultProps, Props, State> {
   }
 
   _renderRouter(props: Props): Object | Component<*, *, *> {
-    const {
-      routes,
-      routerContext,
-      routerHistory,
-    } = props;
-
     // server rendering
-    if (routerContext) return routerContext;
+    if (props.routerContext) return props.routerContext;
 
     // router middleware
     const render: Function = applyRouterMiddleware(
-      useScroll((prevRouterProps, results) => {
+      useScroll((prevRouterProps, { location, routes }) => {
+        // If the user provides custom scroll behaviour, use it, otherwise fallback to the default
+        // behaviour.
+        const { useScroll: customScrollBehavior } = routes.find(route => (
+          route.useScroll &&
+          prevRouterProps &&
+          prevRouterProps.location.pathname !== location.pathname
+        )) || {};
+
+        if (typeof customScrollBehavior === 'function') {
+          return customScrollBehavior(prevRouterProps, location);
+        }
```

This should be a `function` as part of the API. To ensure this use Flow and/or propTypes. We do not need such checkins in the code (neither tests).

---

- Can't use `propTypes`, it's Root component but `useScroll` can be defined on Route component.
- Flow types won't cover it - I've checked - I even changed build process to output `Root.js.flow` file to project so it'll get properly check but still flow doesn't detect error.

---

Still, our API it says it should be a function right? I prefer the code to crash if user passes let's say a boolean that just silently ignore it.

---

Have you seen how helpless are stack traces for some errors? It's better to have that check and add `else` block with meaningful error message - so just by reading it, you would know what you have to fix.